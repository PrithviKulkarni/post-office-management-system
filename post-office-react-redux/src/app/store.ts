import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
