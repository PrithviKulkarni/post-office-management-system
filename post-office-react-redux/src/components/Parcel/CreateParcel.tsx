import React, { SyntheticEvent } from 'react'
import { useNavigate } from 'react-router-dom'
import { useState } from 'react'
import { useAppDispatch } from '../../app/hooks'
import { defaultParcel, Parcel } from '../interfaces/parcelInterface'
import { Container, Button, Form, InputGroup } from 'react-bootstrap'
import { addNewParcel } from '../../features/ParcelSlice'

/***
 * @returns A form to fill by the user to create a new Letter
 */
const CreateParcel = () => {
  const dispatch = useAppDispatch()
  const navigate = useNavigate()
  const [parcel_no, setParcelNo] = useState<number>(0)
  const [parcel_weight, setParcelW] = useState<number>(0)
  const [status, setStatus] = useState<string>('')
  const [estimated_delivery_days, setEDD] = useState<number>(0)

  const [errorMessage, setErrorMessage] = useState<string>('')

  const submitHandler = (event: SyntheticEvent) => {
    console.log('inside SUbmit Handler')
    event.preventDefault()
    

    const validated = [
      parcel_no,
      parcel_weight,
      status,
      estimated_delivery_days
    ].every(Boolean)


    try {
      if (validated) {
        const newParcelData: Parcel = defaultParcel

        if (parcel_no) {
          newParcelData.parcel_no = parcel_no
        }
        if (parcel_weight) {
          newParcelData.parcel_weight = parcel_weight
        }
        if (status) {
          newParcelData.status = status
        }
        if (estimated_delivery_days) {
          newParcelData.estimated_delivery_days = estimated_delivery_days
        }

        dispatch(addNewParcel(newParcelData))
        navigate('/parcels')
       
      }
    } catch (error) {
      setErrorMessage('Parcel NOT FOUND.')
      console.log('errorMessage: ', errorMessage)
    }
  }

  return (
    <div className="createParcel" data-testid="createParcelTest" >
    <Container>
      <h3 className='text-center text-light'> Add a new parcel </h3>

      <Form onSubmit={submitHandler}>
        <InputGroup className='mb-3'>
          <InputGroup.Text id='basic-addon1'>Parcel no</InputGroup.Text>
          <Form.Control
            data-testid="parcelNo-input"
            type='text'
            aria-label='parcel_no'
            aria-describedby='basic-addon1'
            size='sm'
            name='parcel_no'
            value={parcel_no}
            onChange={event => setParcelNo(Number(event.target.value))}
          />
        </InputGroup>

        <InputGroup className='mb-3'>
          <InputGroup.Text id='basic-addon1'>Parcel Weight(kg)</InputGroup.Text>
          <Form.Control
            type='text'
            data-testid= "parcelWeight-input"
            aria-label='parcel_weight'
            aria-describedby='basic-addon1'
            size='sm'
            name='parcel_weight'
            value={parcel_weight}
            onChange={event => setParcelW(Number(event.target.value))}
          />
        </InputGroup>

        <InputGroup className='mb-3'>
          <InputGroup.Text id='basic-addon1'>Status</InputGroup.Text>
          <Form.Control
            type='text'
            data-testid="status-input"
            aria-label='status'
            aria-describedby='basic-addon1'
            size='sm'
            name='status'
            value={status}
            onChange={event => setStatus(event.target.value)}
          />
        </InputGroup>

        <InputGroup className='mb-3'>
          <InputGroup.Text id='basic-addon1'>
            Estimated delivery (in days){' '}
          </InputGroup.Text>
          <Form.Control
            type='text'
            data-testid="edd-input"
            aria-label='estimated_delivery_days'
            aria-describedby='basic-addon1'
            size='sm'
            name='estimated_delivery_days'
            value={estimated_delivery_days}
            onChange={event => setEDD(Number(event.target.value))}
          />
        </InputGroup>

        <Button  type='submit' className="createButton" data-testid="createButton">Add New Parcel</Button>
        <p>{errorMessage}</p>
      </Form>
    </Container>
    </div>
  )
}

export default CreateParcel
