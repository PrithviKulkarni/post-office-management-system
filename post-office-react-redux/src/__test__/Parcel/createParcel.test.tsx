import React from "react";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { store } from "../../app/store";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import CreateParcel from '../../components/Parcel/CreateParcel';

afterEach(cleanup);
it("renders Create Parcel page correctly", () => {
    const{ asFragment } = render(
        <Provider store= {store}>
            <BrowserRouter>
            <CreateParcel/>
            </BrowserRouter>
        </Provider>
    )
    expect(asFragment()).toMatchSnapshot();
});

it("Verifies the Create Parcel render", () => {
   const{getByTestId} =  render(
        <Provider store= {store}>
            <BrowserRouter>
            <CreateParcel/>
            </BrowserRouter>
        </Provider>
    )
    const button = getByTestId("createButton");
    expect(button.textContent).toBe("Add New Parcel");
    
    
    const parcelNo = getByTestId("parcelNo-input")
    const parcelWeight = getByTestId("parcelWeight-input")
    const status = getByTestId("status-input")
    const edd = getByTestId("edd-input")
    fireEvent.change(parcelNo, {target: {value: "1"}});
    fireEvent.change(parcelWeight, {target: {value: "1"}})
    fireEvent.change(status, {target: {value: "test"}})
    fireEvent.change(edd, {target: {value: "1"}} )
    fireEvent.click(button);
    expect(button).toBeTruthy();
    expect(screen.getByTestId("createParcelTest")).toHaveClass("createParcel")
})






