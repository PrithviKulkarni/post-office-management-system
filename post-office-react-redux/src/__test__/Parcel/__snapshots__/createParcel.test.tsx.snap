// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Create Parcel page correctly 1`] = `
<DocumentFragment>
  <div
    class="createParcel"
    data-testid="createParcelTest"
  >
    <div
      class="container"
    >
      <h3
        class="text-center text-light"
      >
         Add a new parcel 
      </h3>
      <form
        class=""
      >
        <div
          class="mb-3 input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Parcel no
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="parcel_no"
            class="form-control form-control-sm"
            data-testid="parcelNo-input"
            name="parcel_no"
            type="text"
            value="0"
          />
        </div>
        <div
          class="mb-3 input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Parcel Weight(kg)
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="parcel_weight"
            class="form-control form-control-sm"
            data-testid="parcelWeight-input"
            name="parcel_weight"
            type="text"
            value="0"
          />
        </div>
        <div
          class="mb-3 input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Status
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="status"
            class="form-control form-control-sm"
            data-testid="status-input"
            name="status"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3 input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Estimated delivery (in days) 
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="estimated_delivery_days"
            class="form-control form-control-sm"
            data-testid="edd-input"
            name="estimated_delivery_days"
            type="text"
            value="0"
          />
        </div>
        <button
          class="createButton btn btn-primary"
          data-testid="createButton"
          type="submit"
        >
          Add New Parcel
        </button>
        <p />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
